patchstore.py

```diff
@@ -0,0 +1,116 @@
+    along with a checksum, and an optional timestamp (in the ISO 8601 format).
+    """
+
+    def __init__(self, dbfile: str):
+        self.dbfile = dbfile
+        logging.info(f'Instantiating Patch Store backed by {dbfile}')
+        if not os.path.exists(dbfile):
+            logging.warn(f'DB file {dbfile} does not exist, creating it')
+        self.db = sqlite3.connect(dbfile)
+        with self.db:
+            self.db.execute(
+                'CREATE TABLE IF NOT EXISTS patches'
+                '(filename TEXT, content BLOB, checksum TEXT, origin TEXT, timestamp TEXT)'
+            )
+
+    def add(self, filename: str, content: str, origin: str, timestamp=None):
+        """
+        Add (insert or update) a patch to the store.
+        """
+        if not filename:
+            logging.error('Storing a patch requires a filename to identify the patch')
+            return False
+        if mimetypes.guess_type(filename)[0] not in ('text/x-diff', 'text/x-patch'):
+            logging.error('Filename does not match that of a patch file')
+            return False
+        if not content:
+            logging.error('Empty patch contents')
+            return False
+        if not origin:
+            logging.error('Storing a patch requires a valid origin')
+            return False
+        if timestamp is not None:
+            try:
+                timestamp = datetime.fromisoformat(timestamp).isoformat(sep=' ')
+            except ValueError:
+                logging.error(f'Invalid timestamp format: {timestamp} (expects ISO 8601)')
+                return False
+        else:
+            timestamp = datetime.now().isoformat(sep=' ')
+
+        encoded_content = base64.b64encode(content.encode())
+        checksum = self.compute_checksum(encoded_content)
+
+        if not self.exists(filename, origin):
+            with self.db:
+                self.db.execute(
+                    'INSERT INTO patches(filename, content, checksum, origin, timestamp) '
+                    'VALUES(?, ?, ?, ?, ?)',
+                    (filename, encoded_content, checksum, origin, timestamp)
+                )
+            logging.info(f'Stored new patch: {filename} ({origin})')
+        else:
+            with self.db:
+                self.db.execute(
+                    'UPDATE patches SET content = ?, checksum = ?, timestamp = ? '
+                    'WHERE filename = ? AND origin = ?',
+                    (encoded_content, checksum, timestamp, filename, origin)
+                )
+            logging.info(f'Updated existing patch: {filename} ({origin})')
+
+        return True
+
+    def compute_checksum(self, encoded_content: bytes):
+        hash = hashlib.sha256(encoded_content)
+        return f'{hash.name}:{hash.hexdigest()}'
+
+    def exists(self, filename: str, origin: str):
+        """
+        Check whether a patch matching a given filename and a given origin
+        already exists in the store.
```

Comment:
```suggestion
        already exists in the store.
        
        :param filename: The filename to search for
        :param origin: The origin identifier of the patch to search
        
        :return: True if the filename/origin exists on the store, False otherwise
```